### ISSUE TITLE
RFC: `usePreventNavigate((url?: URL) => boolean | Promise<boolean>)`

### ISSUE THREAD

**Issue by wmertens** [QwikDev/qwik-evolution#15]
### Champion

@wmertens

### What's the motivation for this proposal?

discussed in #14 

### PRs/ Links / References

https://github.com/QwikDev/qwik/pull/6825

**Comment by thejackshelton** [comment]
So that I understand correctly, if the user was to navigate to this form page, and then fill out a form, let's say they hit the browser `back button`, this would prevent them from going to the previous page?

I sometimes do this on accident and see that the browser sometimes prevents me from going back to the previous URL and I'm relieved. I wonder at what point this would be a frustrating experience.

**Comment by wmertens** [comment]
@thejackshelton yes correct. I hope it's not a frustrating experience when the navigation is only prevented when there are changes. I think that due to it being a hook, it can get at all the metadata it needs to make that decision.

**Comment by AnthonyPAlicea** [comment]
This is terrific functionality, and a common need in web apps when the page in a dirty state. The UX way to avoid frustration is to ensure there is either:

1. An explicit cancelation workflow when usePreventNavigate is used (like a Cancel button next to a Save button), or
2. An auto-save feature, where preventing navigation happens when data has been changed and autosave hasn't yet triggered. 
 
These wouldn't need to be enforced in any way, but could be mentioned in documentation as best practices.

**Comment by shairez** [comment]
thanks @AnthonyPAlicea ! 
Important point!

Now that it's merged and will soon be released as "experimental" we would love a docs PR with your expertise sprinkled all over it 🙏🙌

**Comment by wmertens** [comment]
@shairez what do we need to do to close this?

**Comment by omerman** [comment]
This works great! Brilliant implementation, a spot on hook 🔥
